### PR TITLE
New monotonic

### LINKED ISSUE #247: Task / Resource Handles
**The issue**

Coming from an FreeRTOS background, I am used to being able to pass Queue and task handles around such that none of the tasks need to know of the other tasks existance. I'd like to be able to do the same thing with RTFM.

**An example scenario:**

A Task offloads a HttpRequest to the HttpClient task by enqueing the request onto a queue that the HttpClient Task is listening to. The client performs the request but how does it pass the result back? The result could be a simple message to say it completed, or it could actually have data to return.

One option is having a queue on the Callee task that the client task can push the result onto, but how do we communicate to the client task which queue to send it back to?

I don't really have any strong thoughts on how to solve this (bar hardcoding values, which I am doing currently). I'm guessing there would be some sort of tagging system such that a tag or handle can resolve to a resource in a safe way.

### LINKED ISSUE #292: Populate an enum with the task names
Mainly for use with task callbacks, see: https://github.com/rtfm-rs/cortex-m-rtfm/issues/247#issuecomment-571972834

### LINKED ISSUE #299: Adding priority to scheduled task causes problem with resource methods not found
Simply adding a `priority=2` to a scheduled task causes the rtfm expansion to produce code that will not build.  

I have a repeating scheduled task `imu_reader` that processes a series of incoming messages by calling a method `handle_all_messages`.  If I change the task defintion from:
```
 #[task(resources = [i2c1_driver], schedule = [imu_reader])]
```
to 

```
#[task(resources = [i2c1_driver], schedule = [imu_reader], priority = 2)]
```

I get the `cargo build` error:

```
error[E0599]: no method named `handle_all_messages` found for type `resources::i2c1_driver<'_>` in the current scope
   --> src/main.rs:140:34
    |
39  | #[app(device = stm32h7xx_hal::pac,  peripherals = true, monotonic = rtfm::cyccnt::CYCCNT)]
    | ------------------------------------------------------------------------------------------ method `handle_all_messages` not found for this
...
140 |         cx.resources.i2c1_driver.handle_all_messages();
    |                                  ^^^^^^^^^^^^^^^^^^^ method not found in `resources::i2c1_driver<'_>`
```

I may have misunderstood how to configure the rtfm task, but this is unexpected behavior. 
I've attached the original main.rs file as well as the good and bad expansions. 

[rtfm-priority-problem.zip](https://github.com/rtfm-rs/cortex-m-rtfm/files/4149275/rtfm-priority-problem.zip)

### LINKED ISSUE #309: Monotonic Implementation Difficulties
I am working on a unified time-type library for embedded applications [embedded-time]( https://www.github.com/PTaylor-FluenTech/embedded-time). There is one considerable pain point trying to make it compatible with the Monotonic trait for use with RTFM. That is the requirement that the Duration must implement a special TryInto<u32> to convert into the System Timer ticks. This use of TryInto<u32> feels inappropriate.

What I would like to suggest is instead have the System Timer implement some trait (let's call it `Clock`). This `Clock` trait would require a Fraction/Ratio indicating the precision of the clock. In that way a TryInto<impl Clock> could be implemented to convert a duration to a different precision.

This would be roughly in line with how time is handled in the std::chrono library in C++.

I'm going to look into creating an example implementation of this structure, but any suggestions/comments would be much appreciated.

### LINKED ISSUE #332: Merging SysTick and Monotonic clocks?
Currently, the SysTick timer is used to wake the device when the next scheduled task is due (or SysTick overflows). I understand the desire to make RTIC only reliant on cortex-m resources. I believe it used to be the case that a custom _monotonic_ clock could not be used, but rather the CYCCNT was used exclusively. The _monotonic_ clock was abstracted, so perhaps the _scheduling timer_ (currently SysTick) could also be abstracted.

Another facet of this is that many device peripheral timers/rtcs have a compare register that can generate an interrupt. My current thought is that this would allow one timer to be used both for monotonic and scheduling timer functions.

One possible drawback to this idea is that it may be difficult to continue supporting SysTick as a possible scheduling timer due to its very different functionality.

#331 #312

### LINKED ISSUE #403: Spawning a task from init sometimes takes 30s to run in `0.6.0-alpha.0`
My specs:
- stm32f107 board with a display and 2 buttons connected via gpio
- core clock is 8Mhz
- [current master](https://github.com/rtic-rs/cortex-m-rtic#ad50b54530b7068dde3227dfb72f2a641171af61)
- panic is nop loop

The right button is connected to EXTI to trigger display's backlight using a hardware task (`#[task(binds = EXTI15_10 ...`) and it works fine. The left button is polled every 5ms and it also works when launched from the openocd. 

However, after a power reset w/ no debugger attached, in ~90% of cases it takes *exactly 30 seconds* for the left button to start being polled. In the rest 10% it works fine from the start. I've tried multiple things, but in the end changing 
```rust
poll_left_button::spawn().ok();
``` 
to 
```rust 
poll_left_button::schedule(cx.start).ok();
```

fixed the issue, but I have no idea why! 

Looking at the rtic-expansion.rs, the only suspicious difference between `spawn` and `schedule` is `poll_left_button_INSTANTS`, but I didn't investigate further.

Here are the relevant code parts:
```rust
    const LEFT_BUTTON_POLL_PERIOD_MS: u32 = 5;
    #[init]
    fn init(mut cx: init::Context) -> init::LateResources {
        // Setup clock
        cx.core.DCB.enable_trace();
        cx.core.DWT.enable_cycle_counter();

        let mut flash = cx.device.FLASH.constrain();
        let mut rcc = cx.device.RCC.constrain();
        let clocks = rcc.cfgr.freeze(&mut flash.acr);
        let sysclock = clocks.sysclk();

        // Setup screen backlight control
        let mut gpiod = cx.device.GPIOD.split(&mut rcc.apb2);
        let screen_backlight = gpiod.pd7.into_push_pull_output(&mut gpiod.crl);

        // We need to get EXTI here to work with external interrupts
        let mut gpioe = cx.device.GPIOE.split(&mut rcc.apb2);

        // Setup left button and spawn debounce task for it
        let left_button = gpioe.pe9.into_pull_down_input(&mut gpioe.crh);
        let left_button_poll_period =
            Duration::from_cycles((sysclock / 1000 * LEFT_BUTTON_POLL_PERIOD_MS).0);

        // poll_left_button::spawn().ok(); // takes 30s to start when debugger isn't attached most of the time!
        poll_left_button::schedule(cx.start).ok();
        // etc...
        }


    #[task(
        resources = [left_button,
                     left_button_state,
                     left_button_poll_period])]
    fn poll_left_button(cx: poll_left_button::Context) {
        let pressed: bool = cx.resources.left_button.is_high().unwrap_or(false);
        let edge = cx.resources.left_button_state.update(pressed);
        if edge == Some(debouncr::Edge::Rising) {
            toggle_backlight::spawn(true).ok();
        } else if edge == Some(debouncr::Edge::Falling) {
            toggle_backlight::spawn(false).ok();
        }

        poll_left_button::schedule(cx.scheduled + *cx.resources.left_button_poll_period).ok();
    }

    #[task(resources = [screen_backlight], priority = 3)]
    fn toggle_backlight(cx: toggle_backlight::Context, enabled: bool) {
        if enabled {
            cx.resources.screen_backlight.set_high().ok();
        } else {
            cx.resources.screen_backlight.set_low().ok();
        }
    }

```

p.s. I'm new to embedded so it might be something obvious! 🙂

### LINKED ISSUE #426: `Montotonic::reset()` is not called at the end of `init` in 0.6.0-alpha.0
Minimal example:

```rust
#![no_main]
#![no_std]

use panic_halt as _;
use rtic::app;

// const PERIOD_1S: u32 = 80_000_000;

#[app(device = stm32l4xx_hal::stm32,
      monotonic = rtic::cyccnt::CYCCNT,
      dispatchers = [DFSDM1, DFSDM2])]
mod app {
    use rtic::cyccnt::U32Ext;
    #[resources]
    struct Resources {}

    #[init]
    fn init(_cx: init::Context) -> init::LateResources {
        task1::spawn().ok();

        init::LateResources {}
    }

    #[task]
    fn task1(_cx: task1::Context) {
        task2::schedule(_cx.scheduled + 100.cycles()).ok();
    }

    #[task]
    fn task2(_cx: task2::Context) {
        task1::schedule(_cx.scheduled + 100.cycles()).ok();
    }
}
```

### LINKED ISSUE #427: `cx.scheduled` is not properly initialized in 0.6.0-alpha.0
With the following minimal example, the following log is generated.
The `cx.scheduled` parameter is not initialized to the correct value it seems.

```rust
#![no_main]
#![no_std]

use panic_halt as _;
use rtic::app;

#[app(device = stm32l4xx_hal::stm32,
      monotonic = rtic::cyccnt::CYCCNT,
      dispatchers = [DFSDM1, DFSDM2])]
mod app {
    use rtic::cyccnt::{Instant, U32Ext};
    use rtt_target::{rprintln, rtt_init_print};
    use cortex_m::asm;

    #[resources]
    struct Resources {}

    #[init]
    fn init(mut cx: init::Context) -> init::LateResources {
        cx.core.DCB.enable_trace();
        cx.core.DWT.enable_cycle_counter();

        rtt_init_print!();
        rprintln!("init");

        task1::spawn().ok();

        unsafe { <rtic::cyccnt::CYCCNT as rtic::Monotonic>::reset() };
        init::LateResources {}
    }

    #[idle]
    fn idle(_cx: idle::Context) -> ! {
        rprintln!("idle");
        loop {
            asm::nop();
        }
    }

    #[task]
    fn task1(_cx: task1::Context) {
        rprintln!("task1: {:?}, scheduled: {:?}", Instant::now(), _cx.scheduled);
        task1::schedule(_cx.scheduled + 2_000_000.cycles()).ok();
    }
}
```

```
13:56:24.814 init
13:56:24.814 task1: Instant(106), scheduled: Instant(4027851522)
13:56:24.814 task1: Instant(3482), scheduled: Instant(4029851522)
13:56:24.814 task1: Instant(6867), scheduled: Instant(4031851522)
13:56:24.814 task1: Instant(10252), scheduled: Instant(4033851522)
13:56:24.814 task1: Instant(13661), scheduled: Instant(4035851522)
13:56:24.814 task1: Instant(17070), scheduled: Instant(4037851522)
13:56:24.814 task1: Instant(20480), scheduled: Instant(4039851522)
13:56:24.814 task1: Instant(23890), scheduled: Instant(4041851522)
13:56:24.814 task1: Instant(27299), scheduled: Instant(4043851522)
13:56:24.837 task1: Instant(30708), scheduled: Instant(4045851522)
13:56:24.837 task1: Instant(34117), scheduled: Instant(4047851522)
13:56:24.837 task1: Instant(37526), scheduled: Instant(4049851522)
13:56:24.837 task1: Instant(40935), scheduled: Instant(4051851522)
13:56:24.837 task1: Instant(44344), scheduled: Instant(4053851522)
13:56:24.837 task1: Instant(47753), scheduled: Instant(4055851522)
13:56:24.837 task1: Instant(51162), scheduled: Instant(4057851522)
13:56:24.837 task1: Instant(54571), scheduled: Instant(4059851522)
13:56:24.837 task1: Instant(57980), scheduled: Instant(4061851522)
13:56:24.837 task1: Instant(97030), scheduled: Instant(4085851522)
13:56:25.082 task1: Instant(100502), scheduled: Instant(4087851522)
13:56:25.082 task1: Instant(103818), scheduled: Instant(4089851522)
13:56:25.082 task1: Instant(107134), scheduled: Instant(4091851522)
13:56:25.082 task1: Instant(110450), scheduled: Instant(4093851522)
13:56:25.082 task1: Instant(113766), scheduled: Instant(4095851522)
13:56:25.082 task1: Instant(117082), scheduled: Instant(4097851522)
13:56:25.082 task1: Instant(120398), scheduled: Instant(4099851522)
13:56:25.082 task1: Instant(123714), scheduled: Instant(4101851522)
13:56:25.082 task1: Instant(190724), scheduled: Instant(4145851522)
13:56:25.082 task1: Instant(194097), scheduled: Instant(4147851522)
13:56:25.082 task1: Instant(197492), scheduled: Instant(4149851522)
13:56:25.082 task1: Instant(200887), scheduled: Instant(4151851522)
13:56:25.082 task1: Instant(204282), scheduled: Instant(4153851522)
13:56:25.082 task1: Instant(207677), scheduled: Instant(4155851522)
13:56:25.082 task1: Instant(211072), scheduled: Instant(4157851522)
13:56:25.082 task1: Instant(214467), scheduled: Instant(4159851522)
13:56:25.082 task1: Instant(217862), scheduled: Instant(4161851522)
13:56:25.082 task1: Instant(221257), scheduled: Instant(4163851522)
13:56:25.082 idle
13:56:25.552 task1: Instant(2884483), scheduled: Instant(2884226)
13:56:26.050 task1: Instant(4884483), scheduled: Instant(4884226)
13:56:26.551 task1: Instant(6884483), scheduled: Instant(6884226)
```

### LINKED ISSUE #432: Shorten the path for UI-tests, remove "/single/"
With the removal of multi-core there are no other UI-tests than single core tests.

### LINKED ISSUE #433: Code Gen Panics When there is `monotonic` is defined but no task defined
If an app is defined with `monotonic`
```
#[rtic::app(... , monotonic = rtic::cyccnt::CYCCNT,...)]
```
and no task is defined, the procedural macro fails with 
``` log
help: message: called `Option::unwrap()` on a `None` value
```
on this line:
https://github.com/rtic-rs/cortex-m-rtic/blob/e65a86f6962218b66a14b7ccc40eec7976880eba/macros/src/codegen/pre_init.rs#L79

This is a niche issue but could be encountered when creating new projects. There should be a proper error message for this.

### PR DESCRIPTION
**Pull request by korken89**
Design document: https://hackmd.io/vWa9GvssR8qBfUYgMZm0CQ

Closes #433 
Closes #432
Closes #427
Closes #426 
Closes #403
Closes #332
Closes #312 
Closes #309 
Closes #299 
Closes #292
Closes #247
Closes #219

### CODE REVIEW

**Comment by korken89** [comment]
Needs fixing:

* The codegen assumes only normal interrupt vectors, if the user provides `SysTick` we need to change the interrupt enable/disable code to use `cortex-m` crate.
* Probably more to come

File: macros/src/codegen.rs [hunk 1]
```diff
@@ -106,11 +104,62 @@ pub fn app(app: &App, analysis: &Analysis, extra: &Extra) -> TokenStream2 {
...
+            quote! {
+                pub use rtic::Monotonic as _;
+
+                #[doc(hidden)]
+                pub type #mangled_name = #ty;
+
+                #[allow(non_snake_case)]
+                pub mod #name {
+                    /// Access the global `Monotonic` implementation, not that this will panic
+                    /// before the this `Monotonic` has been passed to the RTIC runtime.
```
Comment (AfoHT) [inline macros/src/codegen.rs:133]: This is a bit unclear

File: macros/src/codegen/module.rs [hunk 2]
```diff
@@ -67,29 +54,11 @@ pub fn codegen(
...
-                fields.push(quote!(
-                    /// Time at which this handler started executing
-                    pub start: <#m as rtic::Monotonic>::Instant
-                ));
-
-                values.push(quote!(start: instant));
-
-                needs_instant = true;
-            }
+            // None for now.
```
Comment (AfoHT) [inline macros/src/codegen/module.rs:57]: Is this a TODO?

File: macros/src/codegen/module.rs [hunk 3]
```diff
@@ -67,29 +54,11 @@ pub fn codegen(
...
-                fields.push(quote!(
-                    /// The time at which this task was scheduled to run
-                    pub scheduled: <#m as rtic::Monotonic>::Instant
-                ));
-
-                values.push(quote!(scheduled: instant));
-
-                needs_instant = true;
-            }
+            // None for now.
```
Comment (AfoHT) [inline macros/src/codegen/module.rs:61]: Is this a TODO?

File: macros/src/lib.rs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#![deny(warnings)]
+// #![deny(warnings)]
```
Comment (AfoHT) [inline macros/src/lib.rs:1]: TODO

File: src/lib.rs [hunk 5]
```diff
@@ -32,126 +32,19 @@
 #![deny(missing_docs)]
 #![deny(rust_2018_compatibility)]
 #![deny(rust_2018_idioms)]
-#![deny(warnings)]
+// #![deny(warnings)]
```
Comment (AfoHT) [inline src/lib.rs:35]: TODO

**Review by AfoHT — CHANGES_REQUESTED** [review]
:+1:  This will be nice

File: macros/src/codegen/module.rs (hunk 2 above)
Comment (korken89) [reply macros/src/codegen/module.rs:57]: This is just to copy the `SoftwareTask` match arm when there is nothing to be done there.

**Review by AfoHT — COMMENTED** [review]
Nice!

**Review by AfoHT — APPROVED** [review]
bors merge

**Pull request merged**